[Usage] Clarify and Update Argument for Specifying Model Revisions

Etelis opened the pull request:
This pull request addresses the need to specify a specific branch or version for various components (weights, code, tokenizer) of a model from Hugging Face when downloading or using a model with the VLLM project. The primary changes include renaming the --revision argument to --weights-revision to avoid ambiguity and updating documentation to ensure consistency and clarity.


FIX #5415

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

Etelis committed:
Update revision naming in config.py

- Renamed 'revision' to 'weights_revision' for clarity.
- Retained the 'revision' naming for the 'from_pretrained' parameter as it appropriately describes its purpose.

Etelis committed:
Rename 'revision' to 'weights-revision' in config.py

- Renamed all instances of 'revision' to 'weights-revision' for consistency.
- Updated ModelConfig arguments to reflect this change.
- Revised the help documentation to use 'weights-revision' for improved clarity.

Etelis committed:
Renaming revisions into weights_revision

-- Renaming the revision to stay consistent with model_config.

Etelis committed:
Rename revision in weight_utils.py

Etelis committed:
Renaming revision to weights_revision in llm.py engine.

Etelis committed:
Renaming params in tests.

Etelis committed:
Renaming revision to weights_revision in AsyncLLM

DarkLight1337 commented on vllm/engine/arg_utils.py:
```diff
@@ -173,12 +173,12 @@ def add_cli_args(
             help='Skip initialization of tokenizer and detokenizer')
         parser.add_argument(
-            '--revision',
+            '--weights-revision',
```
You can use `vllm.utils.deprecate_kwargs` for this.

DarkLight1337 commented on vllm/entrypoints/llm.py:
```diff
@@ -108,7 +109,7 @@ def __init__(
         dtype: str = "auto",
         quantization: Optional[str] = None,
-        revision: Optional[str] = None,
+        weights_revision: Optional[str] = None,
```
Actually, this should not be necessary here as `revision` would be forwarded to `EngineArgs` anyway via `**kwargs`.

mgoin reviewed (commented):
Shouldn't we try to maintain a HF-like interface if possible? `revision` is a fairly visible argument in transformers, so I feel like keeping that interface similar is important, if it doesn't cause many issues.

Etelis committed:
Incorporating deprecation decorator on the EngineArgs class.
-- Re-added the --revision param.
-- Added decorator to EngineArgs for the deprecation of revision
-- When initialising ModelConfig allow the usage of revision still.

Etelis committed:
Merge remote-tracking branch 'origin/main' into usage-fix-5415-specify-hf-branch-download

DarkLight1337 commented on vllm/engine/arg_utils.py:
```diff
@@ -3,24 +3,30 @@
 
-
+@deprecate_kwargs("revision",
+                is_deprecated=lambda: EngineArgs.DEPRECATE_LEGACY,
```
Let's activate the deprecation now.

Etelis replied:
So if I deprecate it now, I will also remove the revision I have re-added. 
Sounds reasonable?

DarkLight1337 replied:
Let's see how the others think about this first.

Etelis committed:
Formating

robertgshaw2-neuralmagic commented:
> > Shouldn't we try to maintain a HF-like interface if possible? `revision` is a fairly visible argument in transformers, so I feel like keeping that interface similar is important, if it doesn't cause many issues.
> 
> As mentioned in the linked issue for this PR, it is confusing that `revision` only applies to the pretrained model config but not tokenizer config. HF uses `revision` in both configs so we need an additional prefix to distinguish between them.

In what situation would we want to allow a different revision for the tokenizer vs the model ... this feels like a footgun waiting to happen

simon-mo commented:
>  HF uses revision in both configs

Can we not deprecate the args and be fully compatible with hf semantics?

DarkLight1337 commented:
> > > Shouldn't we try to maintain a HF-like interface if possible? `revision` is a fairly visible argument in transformers, so I feel like keeping that interface similar is important, if it doesn't cause many issues.
> > 
> > 
> > As mentioned in the linked issue for this PR, it is confusing that `revision` only applies to the pretrained model config but not tokenizer config. HF uses `revision` in both configs so we need an additional prefix to distinguish between them.
> 
> In what situation would we want to allow a different revision for the tokenizer vs the model ... this feels like a footgun waiting to happen

In that case, why do we currently have `revision` for the tokenizer specifically?


> > HF uses revision in both configs
> 
> Can we not deprecate the args and be fully compatible with hf semantics?

The plan is to reintroduce `--revision` to set the version for all components (model weights, tokenizer, and image processor` in another PR.

Etelis commented:
> > > Shouldn't we try to maintain a HF-like interface if possible? `revision` is a fairly visible argument in transformers, so I feel like keeping that interface similar is important, if it doesn't cause many issues.
> > 
> > 
> > As mentioned in the linked issue for this PR, it is confusing that `revision` only applies to the pretrained model config but not tokenizer config. HF uses `revision` in both configs so we need an additional prefix to distinguish between them.
> 
> In what situation would we want to allow a different revision for the tokenizer vs the model ... this feels like a footgun waiting to happen

The current implementation states that if the `--tokenizer-revision` is not set it will be set according to the `--revision`. 
We will re-introduce it with the modified `--revision` I feel like it will be fully backward compatible.

Etelis committed:
Refactor `EngineArgs.from_cli_args` to handle deprecated arguments with `deprecate_kwargs` decorator

- Added `deprecate_kwargs` decorator function to handle deprecated keyword arguments and issue warnings.
- Refactored the `from_cli_args` class method in `EngineArgs` to include an inner function `set_attributes`:
  - `set_attributes` is decorated with `@deprecate_kwargs` to check for deprecated arguments and provide guidance on the new usage.
- Created `get_kwargs_from_args` function to extract argument values from `argparse.Namespace` and convert them to a dictionary.
- Integrated the new inner function within `from_cli_args` to initialize `EngineArgs` instances with deprecation handling.
- Ensured the deprecated `revision` argument triggers a warning and suggests using `weights_revision` instead.

Etelis committed:
Fixed missing arguments revisions in loader.py related to ModelConfig,

Etelis committed:
loader.py additional revision to ModelConfig arguments.

Etelis committed:
Formatting (format.sh)

Etelis commented:
Any update on this?
@DarkLight1337

DarkLight1337 commented:
@simon-mo any comment on my response?

simon-mo commented:
I would prefer making the change in one PR so that it is semantically correct and support `--revision` as is.

simon-mo commented:
for other comments, as long as it is clearly documented in code and docs I'm supportive!
